File: src/components/pagination/pagination.ts
/// <reference path="../../../typings/tsd.d.ts" />

import {
  Component, View, Directive,
  LifecycleEvent, EventEmitter,
  ElementRef,
  DefaultValueAccessor,
  coreDirectives, CSSClass,
  Self, NgModel, Renderer,
  ViewEncapsulation, ViewRef,
  ViewContainerRef, TemplateRef, NgFor, ComponentRef
} from 'angular2/angular2';

// todo: extract base functionality classes
// todo: use lodash#default for configuration
// todo: expose an option to change default configuration
// todo: solve problem with .pagination-sm>li:first-child>a and <template/> from ng-if >.<

const paginationConfig = {
  maxSize: undefined,
  itemsPerPage: 10,
  boundaryLinks: false,
  directionLinks: true,
  firstText: 'First',
  previousText: 'Previous',
  nextText: 'Next',
  lastText: 'Last',
  rotate: true
};

@Component({
  selector: 'pagination, [pagination]',
  properties: [
    'maxSize',
    'rotate',
    'boundaryLinks',
    'totalItems',
    'firstText',
    'previousText',
    'nextText',
    'lastText',
    'disabled',
    'directionLinks',
    'itemsPerPage'
  ],
  events: ['numPages'],
  lifecycle: [LifecycleEvent.onInit]
})
@View({
  template: `
  <ul class="pagination" [class]="classMap">
    <li [class]="{disabled: noPrevious()||disabled, hidden: !boundaryLinks}" class="pagination-first">
      <a href (click)="selectPage(1, $event)">{{getText('first')}}</a></li>

    <li [class]="{disabled: noPrevious()||disabled, hidden: !directionLinks}" class="pagination-prev"><a href (click)="selectPage(page - 1, $event)">{{getText('previous')}}</a></li>

    <li *ng-for="#page of pages" [class]="{active: page.active, disabled: disabled&&!page.active}" class="pagination-page"><a href (click)="selectPage(page.number, $event)">{{page.text}}</a></li>

    <li [class]="{disabled: noNext()||disabled, hidden: !directionLinks}" class="pagination-next"><a href (click)="selectPage(page + 1, $event)">{{getText('next')}}</a></li>

    <li [class]="{disabled: noNext()||disabled, hidden: !boundaryLinks}" class="pagination-last"><a href (click)="selectPage(totalPages, $event)">{{getText('last')}}</a></li>
  </ul>
  `,
  directives: [coreDirectives, CSSClass],
  encapsulation: ViewEncapsulation.NONE
})
export class Pagination extends DefaultValueAccessor {
  public config:any;

  private classMap: string;
  private maxSize:number;
  private rotate:boolean;
  private boundaryLinks:any;

  // labels
  private firstText:string;
  private previousText:string;
  private nextText:string;
  private lastText:string;

  private disabled:boolean;
  private directionLinks:boolean;
  private numPages:EventEmitter = new EventEmitter();

  private _itemsPerPage:number;
  private _totalItems:number;
  private _totalPages:number;

  private get itemsPerPage() {
    return this._itemsPerPage;
  }

  private set itemsPerPage(v:number) {
    this._itemsPerPage = v;
    this.totalPages = this.calculateTotalPages();
  }

  private get totalItems() {
    return this._totalItems;
  }

  private set totalItems(v:number) {
    this._totalItems = v;
    this.totalPages = this.calculateTotalPages();
  }

  private get totalPages() {
    return this._totalPages;
  }

  private set totalPages(v: number) {
    this._totalPages = v;
    console.log(`num pages: ${v}`);
    this.numPages.next(v);
    if (this.page  > v) {
      this.selectPage(v);
    }
  }

  // ??
  private page:number;
  private pages:Array<any>;

  constructor(@Self() cd:NgModel, renderer:Renderer, elementRef:ElementRef) {
    super(cd, renderer, elementRef);
    this.config = this.config || paginationConfig;
  }

  onInit() {
    this.classMap = this.elementRef.nativeElement.getAttribute('class') || '';
    // watch for maxSize
    this.maxSize = typeof this.maxSize !== 'undefined' ? this.maxSize : paginationConfig.maxSize;
    this.rotate = typeof this.rotate !== 'undefined' ? this.rotate : paginationConfig.rotate;
    this.boundaryLinks = typeof this.boundaryLinks !== 'undefined' ? this.boundaryLinks : paginationConfig.boundaryLinks;
    this.directionLinks = typeof this.directionLinks !== 'undefined' ? this.directionLinks : paginationConfig.directionLinks;


    // base class
    this.page = typeof this.page !== 'undefined' ? this.page : 1;
    this.itemsPerPage = typeof this.itemsPerPage !== 'undefined' ? this.itemsPerPage : paginationConfig.itemsPerPage;

    this.totalPages = this.calculateTotalPages();
    // this class
    this.pages = this.getPages(this.page, this.totalPages);
  }

  writeValue(value:number) {
    this.page = value || 1;
    this.pages = this.getPages(this.page, this.totalPages);
  }

  private selectPage(page:number, event?:MouseEvent) {
    if (event) {
      event.preventDefault();
    }

    if ((!this.disabled || !event) && this.page !== page && page > 0 && page <= this.totalPages) {
      event.target.blur();
      this.writeValue(page);
      this.cd.viewToModelUpdate(page);
    }
  }

  private getText(key:string) {
    return this[key + 'Text'] || paginationConfig[key + 'Text'];
  }

  private noPrevious() {
    return this.page === 1;
  }

  private noNext() {
    return this.page === this.totalPages;
  }

  // Create page object used in template
  private makePage(number, text, isActive) {
    return {
      number: number,
      text: text,
      active: isActive
    };
  }

  private getPages(currentPage, totalPages) {
    let pages = [];

    // Default page limits
    let startPage = 1;
    let endPage = totalPages;
    let isMaxSized = typeof this.maxSize !== 'undefined' && this.maxSize < totalPages;

    // recompute if maxSize
    if (isMaxSized) {
      if (this.rotate) {
        // Current page is displayed in the middle of the visible ones
        startPage = Math.max(currentPage - Math.floor(this.maxSize / 2), 1);
        endPage = startPage + this.maxSize - 1;

        // Adjust if limit is exceeded
        if (endPage > totalPages) {
          endPage = totalPages;
          startPage = endPage - this.maxSize + 1;
        }
      } else {
        // Visible pages are paginated with maxSize
        startPage = ((Math.ceil(currentPage / this.maxSize) - 1) * this.maxSize) + 1;

        // Adjust last page if limit is exceeded
        endPage = Math.min(startPage + this.maxSize - 1, totalPages);
      }
    }

    // Add page number links
    for (var number = startPage; number <= endPage; number++) {
      let page = this.makePage(number, number, number === currentPage);
      pages.push(page);
    }

    // Add links to move between page sets
    if (isMaxSized && !this.rotate) {
      if (startPage > 1) {
        let previousPageSet = this.makePage(startPage - 1, '...', false);
        pages.unshift(previousPageSet);
      }

      if (endPage < totalPages) {
        let nextPageSet = this.makePage(endPage + 1, '...', false);
        pages.push(nextPageSet);
      }
    }

    return pages;
  }

  // base class
  private calculateTotalPages() {
    let totalPages = this.itemsPerPage < 1 ? 1 : Math.ceil(this.totalItems / this.itemsPerPage);
    return Math.max(totalPages || 0, 1);
  }
}


const pagerConfig = {
  itemsPerPage: 10,
  previousText: '« Previous',
  nextText: 'Next »',
  align: true
};

@Component({
  selector: 'pager, [pager]',
  properties: [
    'align', 'totalItems', 'previousText', 'nextText', 'itemsPerPage'
  ],
  lifecycle: [LifecycleEvent.onInit]
})
@View({
  template: `
    <ul class="pager">
      <li [class]="{disabled: noPrevious(), previous: align}"><a href (click)="selectPage(page - 1, $event)">{{getText('previous')}}</a></li>
      <li [class]="{disabled: noNext(), next: align}"><a href (click)="selectPage(page + 1, $event)">{{getText('next')}}</a></li>
  </ul>
  `,
  directives: [CSSClass]
})
export class Pager extends Pagination {
  constructor(@Self() cd:NgModel, renderer:Renderer, elementRef:ElementRef) {
    this.config = pagerConfig;
    super(cd, renderer, elementRef);
  }
}

export const pagination = [Pagination, Pager];

File: lib/pagination/pagination.dart
import "package:angular2/angular2.dart";
import 'dart:html';
import 'dart:math';
import 'package:node_shims/js.dart';
// todo: extract base functionality classes

// todo: expose an option to change default configuration

// todo: solve problem with .pagination-sm>li:first-child>a and <template/> from ng-if >.<
const _PAGINATION_CONFIG = const {
  "maxSize" : null,
  "itemsPerPage" : 10,
  "boundaryLinks" : false,
  "directionLinks" : true,
  "firstText" : "First",
  "previousText" : "Previous",
  "nextText" : "Next",
  "lastText" : "Last",
  "rotate" : true
};

@Component (selector: "pagination[ng-model], [pagination][ng-model]",
    inputs: const [
      "rotate",
      "disabled",
      "totalItems",
      "itemsPerPage",
      "maxSize",
      "boundaryLinks",
      "directionLinks",
      "firstText",
      "previousText",
      "nextText",
      "lastText"
    ],
    outputs: const ["numPages"])
@View (template: '''
  <ul class="pagination" [ng-class]="classMap">
    <li class="pagination-first"
        [ng-class]="{disabled: noPrevious()||disabled, hidden: !boundaryLinks}"
        [hidden]="!boundaryLinks">
      <a href (click)="selectPage(1, \$event)">{{firstText}}</a>
    </li>

    <li class="pagination-prev"
        [ng-class]="{disabled: noPrevious()||disabled, hidden: !directionLinks}"
        [hidden]="!directionLinks">
      <a href (click)="selectPage(page - 1, \$event)">{{previousText}}</a>
      </li>

    <li *ng-for="#page of pages" [ng-class]="{active: page['active'], disabled: disabled && !page['active']}" class="pagination-page">
      <a href (click)="selectPage(page['number'], \$event)">{{page['text']}}</a>
    </li>

    <li class="pagination-next"
        [ng-class]="{disabled: noNext()||disabled, hidden: !directionLinks}"
        [hidden]="!directionLinks">
      <a href (click)="selectPage(page + 1, \$event)">{{nextText}}</a></li>

    <li class="pagination-last"
        [ng-class]="{disabled: noNext()||disabled, hidden: !boundaryLinks}"
        [hidden]="!boundaryLinks">
      <a href (click)="selectPage(totalPages, \$event)">{{lastText}}</a></li>
  </ul>
  ''',
    directives: const [ CORE_DIRECTIVES, NgClass],
    encapsulation: ViewEncapsulation.None)
class Pagination extends DefaultValueAccessor implements OnInit {
  Pagination(this.ngModel, Renderer renderer, ElementRef elementRef)
      : super (renderer, elementRef) {
    this.elementRef = elementRef;
    ngModel.valueAccessor = this;
  }

  NgModel ngModel;

  Map config;

  String classMap;

  num maxSize;

  bool rotate = true;

  dynamic boundaryLinks = false;

  // labels
  String firstText = "First";

  String previousText = "Previous";

  String nextText = "Next";

  String lastText = "Last";

  bool disabled = false;

  bool directionLinks = true;

  EventEmitter numPages = new EventEmitter ();

  int _itemsPerPage = 10;

  int _totalItems = 10;

  int _totalPages = 10;

  get itemsPerPage => _itemsPerPage;

  set itemsPerPage(int v) {
    _itemsPerPage = v;
    totalPages = calculateTotalPages();
  }

  int get totalItems => _totalItems;

  set totalItems(int v) {
    _totalItems = v;
    totalPages = calculateTotalPages();
  }

  get totalPages {
    return _totalPages;
  }

  set totalPages(int v) {
    _totalPages = v;
    numPages.add(v);
    if (page > v) {
      selectPage(v);
    }
  }

  // ??
  int page = 1;
//  int get page => _page;
//  set page(int page) {
//    _page = page ?? 1;
//    pages = getPages(_page, totalPages);
//    pageEmitter.add(_page);
//  }

  List<Map> pages = [];

  ElementRef elementRef;

  onInit() {
    classMap = elementRef.nativeElement.getAttribute("class") ?? "";
    totalPages = calculateTotalPages();
    pages = getPages(page, totalPages);
  }

  writeValue(num value) {
    page = value ?? 1;
    pages = getPages(page, totalPages);
  }

  selectPage(num _page, [MouseEvent event]) {
    if (event != null) {
      event.preventDefault();
    }
    if ((!disabled || event == null) && this.page != _page && _page > 0 &&
        _page <= totalPages) {
      dynamic target = event.target;
      target.blur();
      page = _page;
      pages = getPages(_page, totalPages);
      ngModel.viewToModelUpdate(_page);
    }
  }

  noPrevious() => page <= 1;

  noNext() => page >= totalPages;

  // Create page object used in template
  makePage(number, text, isActive) {
    return { "number" : number, "text" : text, "active" : isActive};
  }

  getPages(currentPage, totalPages) {
    var pages = [];
    // Default page limits
    var startPage = 1;
    var endPage = totalPages;
    var isMaxSized = maxSize != null && maxSize < totalPages;
    // recompute if maxSize
    if (isMaxSized) {
      if (rotate) {
        // Current page is displayed in the middle of the visible ones
        startPage = max(currentPage - (maxSize / 2).floor(), 1);
        endPage = startPage + maxSize - 1;
        // Adjust if limit is exceeded
        if (endPage > totalPages) {
          endPage = totalPages;
          startPage = endPage - maxSize + 1;
        }
      } else {
        // Visible pages are paginated with maxSize
        startPage = (((currentPage / maxSize).ceil() - 1) * maxSize) + 1;
        // Adjust last page if limit is exceeded
        endPage = min(startPage + maxSize - 1, totalPages);
      }
    }
    // Add page number links
    for (var number = startPage; number <= endPage; number ++) {
      var page = makePage(number, number, number == currentPage);
      pages.add(page);
    }
    // Add links to move between page sets
    if (isMaxSized && !rotate) {
      if (startPage > 1) {
        var previousPageSet = makePage(startPage - 1, "...", false);
        unshift(pages, previousPageSet);
      }
      if (endPage < totalPages) {
        var nextPageSet = makePage(endPage + 1, "...", false);
        push(pages, nextPageSet);
      }
    }
    return pages;
  }

  // base class
  calculateTotalPages() {
    var totalPages = itemsPerPage < 1 ? 1 : (
        totalItems / itemsPerPage).ceil();
    return max(totalPages ?? 0, 1);
  }
}

@Component (selector: "pager[ng-model], [pager][ng-model]",
    inputs: const [
      "align", "totalItems", "itemsPerPage", "previousText", "nextText"])
@View (template: '''
<ul class="pager">
  <li [ng-class]="{disabled: noPrevious(), previous: align, \'pull-left\': align}"><a href (click)="selectPage(page - 1, \$event)">{{previousText}}</a></li>
  <li [ng-class]="{disabled: noNext(), next: align, \'pull-right\': align}"><a href (click)="selectPage(page + 1, \$event)">{{nextText}}</a></li>
</ul>
''', directives: const [ NgClass])
class Pager extends Pagination
    implements OnInit {
  bool align = true;

  Pager(@Self() NgModel ngModel, Renderer renderer, ElementRef elementRef)
      : super (ngModel, renderer, elementRef) {
    _itemsPerPage = 10;
    previousText = "« Previous";
    nextText = "Next »";
    align = true;
  }
}

const PAGINATION_DIRECTIVES = const [ Pagination, Pager];